lib/fizzy/instantiate.hpp

```diff
@@ -28,7 +28,18 @@ struct ExternalFunction
     FuncType type;
 };
 
-using table_elements = std::vector<std::optional<ExternalFunction>>;
+// Table element, which references a function in any instance.
+struct TableElement
+{
+    Instance* instance;
+    FuncIdx func_idx;
+    // This pointer is empty most of the time and is used only to keep instance alive in one edge
+    // case, when start function traps, but instantiate has already modified some elements of a
+    // shared (imported) table.
+    std::shared_ptr<Instance> shared_instance;
```

Comment:
This is a simple way to keep instance alive as long as there is function using it in some table. This was previously achieved by capturing this `shared_ptr<Instance>` in `std::function` of table element.

Now this looks like an overhead included in each table element every time, but needed only in weird edge case.
But on the other hand, previously there was `std::function`'s dynamic allocation overhead for each table element all the time, and now this `shared_ptr` stays empty in normal cases, so this theoretically should be overall more efficient than before.

Reply:
This is better now. But I'd still prefer to handle that differently (not necessarily in this PR): either separate `start()` from `instantiate()` or return tuble `Instance*, bool` from `instantiate()`.